File: portfolio/projects/D_INSA/INSA-salary-usb/frontend/src/api/evalObjection.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import apiClient from "./client";

export type ObjectionStatus = "PENDING" | "REVIEWED" | "ACCEPTED" | "REJECTED";
export type ReviewDecision = "ACCEPTED" | "REJECTED";

export interface ObjectionResponse {
  id: number;
  emp_id: number;
  round_id: number;
  reason: string;
  status: ObjectionStatus;
  created_at: string;
}

export interface ObjectionCreatePayload {
  round_id: number;
  reason: string;
}

export interface ObjectionReviewPayload {
  objection_id: number;
  decision: ReviewDecision;
  comment?: string | null;
}

export function useObjections(round_id: number | null) {
  return useQuery({
    queryKey: ["eval", "objections", "list", { round_id }],
    queryFn: async () => {
      const { data } = await apiClient.get<ObjectionResponse[]>(
        "/eval/objections",
        { params: { round_id: round_id ?? undefined } },
      );
      return data;
    },
    enabled: round_id != null,
  });
}

export function useCreateObjection() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: ObjectionCreatePayload) => {
      const { data } = await apiClient.post<ObjectionResponse>(
        "/eval/objections",
        body,
      );
      return data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["eval", "objections", "list"] });
    },
  });
}

export function useReviewObjection() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: ObjectionReviewPayload) => {
      const { data } = await apiClient.put(
        `/eval/objections/${payload.objection_id}/review`,
        { decision: payload.decision, comment: payload.comment ?? null },
      );
      return data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["eval", "objections", "list"] });
    },
  });
}

File: portfolio/projects/D_INSA/INSA-salary-usb/backend/app/schemas/eval_objection.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ObjectionStatus = Literal["PENDING", "REVIEWED", "ACCEPTED", "REJECTED"]
ReviewDecision = Literal["ACCEPTED", "REJECTED"]


class ObjectionCreate(BaseModel):
    round_id: int
    reason: str = Field(min_length=1)


class ObjectionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    emp_id: int
    round_id: int
    reason: str
    status: ObjectionStatus
    created_at: datetime


class ObjectionReviewCreate(BaseModel):
    decision: ReviewDecision
    comment: str | None = None


class ObjectionReviewResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    objection_id: int
    reviewer_id: int
    decision: str
    comment: str | None = None
    reviewed_at: datetime
